### ISSUE TITLE
Persist old versions of a pipeline

### ISSUE THREAD

**Issue by gabrielgrant** [pachyderm/pachyderm#1940]
The dashboard shows the current version number of a pipeline. Clicking on this should show a list of all previous pipeline versions, but there's currently no way to directly access old versions of a pipeline -- the closest I can do get the version of a transform that was used in each job, and, by looking through those build up a history of the pipeline. But even that won't be a complete list: if a pipeline is updated without any jobs being created, then the transform spec is permanently lost.

Old pipeline spec versions should be persisted before being overwritten, and the list should be retrievable through the API

**Comment by jdoliner** [comment]
Closing in favor of #1036

**Comment by gabrielgrant** [comment]
Maybe I'm missing something from the discussion on #1036, but reading that it sounds like a lot bigger feature request than what i'm asking for here (and I'm also not sure that it would actually address this issue). Whereas #1036 is asking about how to have multiple versions of a pipeline used/developed in parallel, what I'm asking about is (I believe) a pretty simple tweak to the way things work today: basically where we currently increment the pipeline version number when a new pipeline spec is deployed, I'm just asking to store a snapshot of the whole previous `PipelineInfo` message.

I think this could be as simple as just adding `repeated PipelineInfo pipeline_history = 23;` or whatever into the `PipelineInfo` proto message, then having whatever code currently increments the version number just move the history list over from the previously-stored PipelineInfo and append that previous revision whenever a new pipeline spec is submitted. Maybe there's more complexity I'm not seeing, though?

**Comment by jdoliner** [comment]
#1036 is a bigger feature but the functionality here is a subset if its functionality and it makes more sense to implement them together rather than have 2 different ways to store multiple copies of a pipeline.

> I think this could be as simple as just adding `repeated PipelineInfo pipeline_history = 23;` or whatever into the PipelineInfo

This is a very weird schema, representing different version of an API object recursively. What does `pipelineInfo.PipelineHistory[0].PipelineHistory` represent?
We don't want to just hack in this functionality we want to do it in a coherent way that's going to match with the rest of the API.

**Comment by gabrielgrant** [comment]
I'm certainly not tied to any particular API for this functionality, but I'm still not seeing how #1036 proposes to provide history of the pipeline versions that output to a single branch

(to answer your specific question, though, it doesn't seem that crazy to just have a history list be populated on the top-level `PipelineInfo` message, does it?)

**Comment by jdoliner** [comment]
#1036 requires having multiple versions of the same pipeline so that you can have, for example a production version and a dev version that people are working on. These would be the same pipeline, in that it would have the same name and output to the same repo. But they both exist concurrently within the system. My plan for #1036 would be to first implement having multiple versions of a pipeline stored in etcd with only the most recent one running. That would satisfy this issue I believe, then from there we'd start building the more complicated parts of the workflow like having mutliple running versions and things like that.

**Comment by jdoliner** [comment]
And it does seem a bit weird to me to have a history populated on a top-level `PipelineInfo` it seems a lot better to have the different versions of the pipeline exist at the same level in the api rather than recursively referencing each other.
